File: 2015/d3p2.cpp
// https://adventofcode.com/2015/day/3

#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

int main() {
    std::ifstream ReadFile("input.txt");
    std::string fileInput;
    int robox = 0;
    int roboy = 0;
    int santax = 0;
    int santay = 0;
    bool isSanta = false;
    std::vector<std::string> visits;
    std::string coords;

    visits.push_back("0|0");

    while(getline(ReadFile, fileInput)) {
        for(int i=0; i < fileInput.size(); i++) {

            isSanta = !isSanta;
            
            if((fileInput[i] == '^') && (isSanta == true)) {
                santay += 1;
            } else if((fileInput[i] == 'v') && (isSanta == true)) {
                santay -= 1;
            } else if((fileInput[i] == '>') && (isSanta == true)) {
                santax += 1;
            } else if((fileInput[i] == '<') && (isSanta == true)) {
                santax -= 1;
            } else if((fileInput[i] == '^') && (isSanta == false)) {
                roboy += 1;
            } else if((fileInput[i] == 'v') && (isSanta == false)) {
                roboy -= 1;
            } else if((fileInput[i] == '>') && (isSanta == false)) {
                robox += 1;
            } else if((fileInput[i] == '<') && (isSanta == false)) {
                robox -= 1;
            }

            std::string stringx = std::to_string(santax);
            std::string stringy = std::to_string(santay);
            std::string stringrx = std::to_string(robox);
            std::string stringry = std::to_string(roboy);

            if(isSanta == true) {
                coords = stringx + "|" + stringy/* + "|0"*/;
            } else if(isSanta == false){
                coords = stringrx + "|" + stringry/* + "|1"*/;
            }

            std::cout << coords << "\n";

            if(std::find(visits.begin(), visits.end(), coords) != visits.end()) {
                continue;
            } else {
                visits.push_back(coords);
            }

        }
    }

    std::cout << visits.size() << "\n";

    return 0;
}

File: 2015/Day 3/d3p2.rs
use std::fs;

fn main() {
    let contents = fs::read_to_string("input.txt").expect("Failed");
    let mut coords: Vec<String> = Vec::new();
    let mut santax: i32 = 0;
    let mut santay: i32 = 0;
    let mut robox: i32 = 0;
    let mut roboy: i32 = 0;
    let mut isSanta: bool = false;

    coords.push("0|0".to_string());

    for char in contents.chars() {

        isSanta = !isSanta;
        if char == '^' && isSanta == true {
            santay += 1;
        } else if char == 'v' && isSanta == true {
            santay -= 1;
        } else if char == '>' && isSanta == true{
            santax += 1;
        } else if char == '<' && isSanta == true {
            santax -= 1;
        } else if char == '^' && isSanta == false {
            roboy += 1;
        } else if char == 'v' && isSanta == false {
            roboy -= 1;
        } else if char == '>' && isSanta == false {
            robox += 1;
        } else if char == '<' && isSanta == false {
            robox -= 1;
        }

        if isSanta == true {
            let visit: String = santax.to_string() + "|" + &santay.to_string();
            for element in coords.clone() {
                if element.to_string() == visit {
                    break;
                }

                if element.to_string() == coords.last().expect("Failed").to_string() {
                    coords.push(visit.clone());
                }
            }
        } else {
            let visit: String = robox.to_string() + "|" + &roboy.to_string();
            for element in coords.clone() {
                if element.to_string() == visit {
                    break;
                }

                if element.to_string() == coords.last().expect("failed").to_string() {
                    coords.push(visit.clone());
                }
        }

    }

    println!("{}", coords.len());
}
}
